cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/MessagingProgramStateWriter.java

```diff
@@ -0,0 +1,179 @@
+ */
+
+package co.cask.cdap.internal.app.program;
+
+import co.cask.cdap.api.messaging.TopicNotFoundException;
+import co.cask.cdap.app.runtime.ProgramOptions;
+import co.cask.cdap.app.runtime.ProgramStateWriter;
+import co.cask.cdap.common.ServiceUnavailableException;
+import co.cask.cdap.common.app.RunIds;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.service.RetryStrategies;
+import co.cask.cdap.common.service.RetryStrategy;
+import co.cask.cdap.internal.app.runtime.ProgramOptionConstants;
+import co.cask.cdap.messaging.MessagingService;
+import co.cask.cdap.messaging.client.StoreRequestBuilder;
+import co.cask.cdap.proto.BasicThrowable;
+import co.cask.cdap.proto.Notification;
+import co.cask.cdap.proto.ProgramRunStatus;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.ProgramRunId;
+import co.cask.cdap.proto.id.TopicId;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableMap;
+import com.google.gson.Gson;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+
+/**
+ * An implementation of ProgramStateWriter that publishes program status events to TMS
+ */
+public final class MessagingProgramStateWriter implements ProgramStateWriter {
+  private static final Logger LOG = LoggerFactory.getLogger(MessagingProgramStateWriter.class);
+  private static final Gson GSON = new Gson();
+  private final MessagingService messagingService;
+  private final TopicId topicId;
+  private final RetryStrategy retryStrategy;
+
+  @Inject
+  public MessagingProgramStateWriter(CConfiguration cConf, MessagingService messagingService) {
+    this.topicId = NamespaceId.SYSTEM.topic(cConf.get(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC));
+    this.retryStrategy = RetryStrategies.fromConfiguration(cConf, "program.message.");
+    this.messagingService = messagingService;
+  }
+
+  @Override
+  public void start(ProgramRunId programRunId, ProgramOptions programOptions, @Nullable String twillRunId) {
+    long startTime = RunIds.getTime(RunIds.fromString(programRunId.getRun()), TimeUnit.MILLISECONDS);
+    if (startTime == -1) {
+      // If RunId is not time-based, use current time as start time
+      startTime = System.currentTimeMillis();
+    }
+    ImmutableMap.Builder properties = ImmutableMap.<String, String>builder()
+      .put(ProgramOptionConstants.START_TIME, String.valueOf(startTime))
+      .put(ProgramOptionConstants.PROGRAM_STATUS, ProgramRunStatus.STARTING.name())
+      .put(ProgramOptionConstants.USER_OVERRIDES, GSON.toJson(programOptions.getUserArguments().asMap()))
+      .put(ProgramOptionConstants.SYSTEM_OVERRIDES, GSON.toJson(programOptions.getArguments().asMap()));
+    if (twillRunId != null) {
+      properties.put(ProgramOptionConstants.TWILL_RUN_ID, twillRunId);
```

Comment:
My intellij gives warning for this. does it happen for you as well? When I changed the declaration as `ImmutableMap.Builder<String, String> properties` the warning went away. better to fix it to have no warnings.